tls/s2n_send.c

```diff
@@ -32,6 +32,40 @@
 #include "utils/s2n_safety.h"
 #include "utils/s2n_blob.h"
 
+bool s2n_should_flush(struct s2n_connection *conn, ssize_t total_message_size, uint16_t max_write_size)
+{
+    /* If the connection is unbuffered then flush on every record written.
+     * The rest of this function assumes conn->send_mode == S2N_BUFFERED_SEND */
+    if (conn->send_mode == S2N_UNBUFFERED_SEND) {
+        return true;
+    }
+
+    uint32_t available_space = s2n_stuffer_space_remaining(&conn->out);
+
+    /* If the stuffer can't store the max possible max fragment size 
+     * without growing it's time to flush. */
+    if (available_space < max_write_size) {
+        return true;
+    }
+
+    uint32_t bytes_in_stuffer = s2n_stuffer_data_available(&conn->out);
+
+    /* If the stuffer cannot contain a max fragment size record then it is time to flush.
+     * This is not redundant with the above check if the conn->send_buffer_size is smaller
+     * than the conn->out stuffer. This is possible because the conn->out stuffer MUST 
+     * be growable currently. */
+    if (bytes_in_stuffer + max_write_size > conn->send_buffer_size) {
+        return true;
+    }
```

Comment:
But if this method is correct, even though the send buffer is technically growable, shouldn't it never grow?

This check also makes me nervous because conn->send_buffer_size could be 0. I think that if it is 0 than the previous check should have returned true, but it still seems dangerous to include a check with a possibly deceptive value. What if someone reorders the two not obviously related checks?

You could also make conn->send_buffer_size non-zero and accurate even if in "unbuffered" mode. It'd definitely be safer, but a potentially quite tricky because in "unbuffered" mode it depends on the max fragment length extension and some connection settings https://github.com/aws/s2n-tls/blob/main/api/s2n.h#L1556-L1574 Maybe don't set it until after the clienthello/serverhello instead of when setting the config?

Reply:
> But if this method is correct, even though the send buffer is technically growable, shouldn't it never grow?

It's a defensive check since the out stuffer being growable makes it possible for the out stuffer to be resized to a value larger than `conn->send_buffer_size`. I think that would be okay but wasteful.

> You could also make conn->send_buffer_size non-zero and accurate even if in "unbuffered" mode. It'd definitely be safer, but a potentially quite tricky because in "unbuffered" mode it depends on the max fragment length extension and some connection settings https://github.com/aws/s2n-tls/blob/main/api/s2n.h#L1556-L1574 Maybe don't set it until after the clienthello/serverhello instead of when setting the config? Could also address https://github.com/aws/s2n-tls/pull/3326#discussion_r893892516.

I guess it could be updated to track the allocation size when the stuffer is first created but I'm not sure if that would feel safe, since the out stuffer size is what really should be getting checked. `conn->send_buffer_size` shouldn't be smaller than S2N_TLS_MAXIMUM_RECORD_LENGTH so I could add a defensive check `if conn->send_buffer_size < S2N_TLS_MAXIMUM_RECORD_LENGTH) return true;`

> What if someone reorders the two not obviously related checks?

Yea I agree it is pretty deceptive. I think I could encapsulate them in a separate function but as I mentioned earlier I think trying to accurately track `conn->send_buffer_size` at runtime could be tricky.

Reply:
ACK - That makes sense. I'll add those changes

Reply:
If we don't rename conn->send_buffer_size, then we really should probably calculate it during the handshake rather than either using it or a calculated value during the first send.